File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/remoteconfiguration/ConfigurationCache.java
package com.snowplowanalytics.snowplow.internal.remoteconfiguration;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.configuration.RemoteConfiguration;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;

public class ConfigurationCache {

    @Nullable
    private String cacheFilePath;
    @Nullable
    private FetchedConfigurationBundle configuration;
    @NonNull
    private RemoteConfiguration remoteConfiguration;

    public ConfigurationCache(@NonNull RemoteConfiguration remoteConfiguration) {
        this.remoteConfiguration = remoteConfiguration;
    }

    @Nullable
    public synchronized FetchedConfigurationBundle readCache(@NonNull Context context) {
        if (configuration != null) {
            return configuration;
        }
        loadCache(context);
        return configuration;
    }

    public synchronized void writeCache(@NonNull Context context, @NonNull FetchedConfigurationBundle configuration) {
        this.configuration = configuration;
        storeCache(context, configuration);
    }

    public synchronized void clearCache(@NonNull Context context) {
        String path = getCachePath(context);
        File file = new File(path);
        file.delete();
    }

    // Private methods

    private String getCachePath(@NonNull Context context) {
        if (cacheFilePath != null) {
            return cacheFilePath;
        }
        String cacheDirPath = context.getCacheDir().getAbsolutePath() + File.separator + "snowplow-cache";
        File cacheDir = new File(cacheDirPath);
        if (!cacheDir.exists()) {
            cacheDir.mkdirs();
        }
        String fileName = "remoteConfig-" + remoteConfiguration.endpoint.hashCode() + ".data";
        cacheFilePath = cacheDir.getAbsolutePath() + File.separator + fileName;
        return cacheFilePath;
    }

    private void loadCache(@NonNull Context context) {
        String path = getCachePath(context);
        ObjectInputStream objectIn = null;
        try {
            FileInputStream fileIn = new FileInputStream(path);
            objectIn = new ObjectInputStream(fileIn);
            configuration = (FetchedConfigurationBundle)objectIn.readObject();
        } catch (FileNotFoundException e) {
            // TODO log exception
        } catch (IOException e) {
            // TODO log exception
        } catch (ClassNotFoundException e) {
            // TODO log exception
        } finally {
            if (objectIn != null) {
                try {
                    objectIn.close();
                } catch (IOException e) {}
            }
        }
    }

    private void storeCache(@NonNull Context context, @NonNull FetchedConfigurationBundle configuration) {
        String path = getCachePath(context);
        ObjectOutputStream objectOut = null;
        try {
            FileOutputStream fileOut = new FileOutputStream(path,false);
            objectOut = new ObjectOutputStream(fileOut);
            objectOut.writeObject(configuration);
            fileOut.getFD().sync();
        } catch (IOException e) {
            // TODO log exception
        } finally {
            if (objectOut != null) {
                try {
                    objectOut.close();
                } catch (IOException e) {}
            }
        }
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/core/remoteconfiguration/ConfigurationCache.kt
package com.snowplowanalytics.core.remoteconfiguration

import android.content.Context
import com.snowplowanalytics.snowplow.configuration.RemoteConfiguration
import java.io.*

class ConfigurationCache(private val remoteConfiguration: RemoteConfiguration) {
    private var cacheFilePath: String? = null
    private var configuration: FetchedConfigurationBundle? = null
    
    @Synchronized
    fun readCache(context: Context): FetchedConfigurationBundle? {
        if (configuration != null) { return configuration }
        
        loadCache(context)
        return configuration
    }

    @Synchronized
    fun writeCache(context: Context, configuration: FetchedConfigurationBundle) {
        this.configuration = configuration
        storeCache(context, configuration)
    }

    @Synchronized
    fun clearCache(context: Context) {
        val path = getCachePath(context)
        val file = File(path)
        file.delete()
    }

    // Private methods
    private fun getCachePath(context: Context): String {
        cacheFilePath?.let { return it }
        
        val cacheDirPath = context.cacheDir.absolutePath + File.separator + "snowplow-cache"
        val cacheDir = File(cacheDirPath)
        if (!cacheDir.exists()) {
            cacheDir.mkdirs()
        }
        val fileName = "remoteConfig-" + remoteConfiguration.endpoint.hashCode() + ".data"
        cacheFilePath = cacheDir.absolutePath + File.separator + fileName
        return cacheFilePath!!
    }

    private fun loadCache(context: Context) {
        val path = getCachePath(context)
        var objectIn: ObjectInputStream? = null
        try {
            val fileIn = FileInputStream(path)
            objectIn = ObjectInputStream(fileIn)
            configuration = objectIn.readObject() as? FetchedConfigurationBundle
        } catch (e: FileNotFoundException) {
            // TODO log exception
        } catch (e: IOException) {
            // TODO log exception
        } catch (e: ClassNotFoundException) {
            // TODO log exception
        } finally {
            if (objectIn != null) {
                try {
                    objectIn.close()
                } catch (e: IOException) {
                }
            }
        }
    }

    private fun storeCache(context: Context, configuration: FetchedConfigurationBundle) {
        val path = getCachePath(context)
        var objectOut: ObjectOutputStream? = null
        try {
            val fileOut = FileOutputStream(path, false)
            objectOut = ObjectOutputStream(fileOut)
            objectOut.writeObject(configuration)
            fileOut.fd.sync()
        } catch (e: IOException) {
            // TODO log exception
        } finally {
            if (objectOut != null) {
                try {
                    objectOut.close()
                } catch (e: IOException) {
                }
            }
        }
    }
}
